sw/device/tests/sram_ctrl_execution_test_ret.c

```diff
@@ -0,0 +1,175 @@
+static const uint32_t kRamEndAddr =
+    TOP_EARLGREY_SRAM_CTRL_RET_AON_RAM_BASE_ADDR +
+    TOP_EARLGREY_SRAM_CTRL_RET_AON_RAM_SIZE_BYTES - 1;
+
+/*
+ * Buffer to hold instructions to be executed.
+ *
+ * This can be thought of as the implementation for the `ram_exec_function_t`.
+ */
+static const uint32_t execution_test_instructions[2] = {
+    /*
+     *   Abstract representation: add rd, rs1, rs2
+     *   +-----------+-------+-------+-------+------+-----------+
+     *   |  0000000  |  rs2  |  rs1  |  000  |  rd  |  0110011  |
+     *   +-----------+-------+-------+-------+------+-----------+
+     *   31        25 24   20 19   15 14   12 11   7 6          0
+     *
+     *   Custom representation: add a0, a1, a0
+     *   +-----------+---------+---------+-------+--------+-----------+
+     *   |  0000000  |  01010  |  01011  |  000  | 01010  |  0110011  |
+     *   +-----------+---------+---------+-------+--------+-----------+
+     *   31        25 24     20 19     15 14   12 11     7 6          0
+     */
+    0x00A58533,
+    /*
+     *   Abstract representation: ret == jalr rd, offset(rs1)
+     *   +----------------+-------+-------+------+-----------+
+     *   |  000000000000  |  rs1  |  000  |  rd  |  1100111  |
+     *   +----------------+-------+-------+------+-----------+
+     *   31             20 19   15 14   12 11   7 6          0
+     *
+     *   Custom representation: jalr zero, 0(ra)
+     *   +----------------+---------+-------+---------+-----------+
+     *   |  000000000000  |  00001  |  000  |  00000  |  1100111  |
+     *   +----------------+---------+-------+---------+-----------+
+     *   31             20 19     15 14   12 11      7 6          0
+     */
+    0x00008067,
+};
+
+/*
+ * Handles an exception.
+ *
+ * This exception handler keeps default exception implementations for faults
+ * that are not relevant in terms of this test. The only fault handler that
+ * is overridden is Instruction Access Fault.
+ *
+ * The controlled Instruction Access Fault occurs originates in the retention
+ * SRAM, which means that normally the return address would be calculated
+ * relevant to the offending instruction. However, due to execution from
+ * retention SRAM being permanently disabled, this approach would not work.
+ * The return address for the one stack frame before is used instead. In
+ * other words sram_execution_test -> retention_sram -> exception_handler
+ * -> sram_execution_test.
+ *
+ * Before the jump into the exception handler, the register set is saved on
+ * stack. This means that the return address can be loaded from the stack, or
+ * simply read from the return address register (before it is overridden by
+ * the code in this exception handler). This means that the return address
+ * register value must be read before any function call.
+ */
+void handler_exception(void) {
+  // Save the return address register value.
+  uint32_t ret_addr = 0;
+  asm volatile("sw ra, %0" : "=m"(ret_addr));
```

Comment:
```
asm ("mv ra, %0" : "=r"(ret_addr));
```
Doesn't need to be volatile.

Reply:
Why doesn't it work with `=r`?

Also, note that we don't mark things as volatile unless it is actually necessary.

That said, looking at this, again, this is actually the wrong code to be writing. What you want is `__builtin_return_address(0)`, so the compiler can do this funny business for you. Can you add
```
#define OT_RETURN_ADDR() \
  __builtin_return_address(0)
```
to `macros.h`? This is much likelier to not break than what you've written. (Regardless of volatile and asm constraints, pulling stuff out of the calling convention using inline assembly is a disaster waiting to happen.)

Reply:
Thank you for this, It completely escape my mind to have a look in builtins.

Done.